servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/PublisherBuffer.java

```diff
@@ -190,57 +222,113 @@ public void onComplete() {
 
-        private final int firstItemsRequestN;
+        private static final long NEED_TERMINATE = MIN_VALUE;
+        private static final long NEED_REQUEST_ITEMS = NEED_TERMINATE + 1;
+        private static final long LAST_SPECIAL_STATE = NEED_REQUEST_ITEMS;  // update this if a new state is added above
+        private static final AtomicLongFieldUpdater<State> pendingUpdater =
+                AtomicLongFieldUpdater.newUpdater(State.class, "pending");
+
         /**
          * Following values are assigned to this variable:
          * <ul>
          *     <li>{@code null} till the first accumulator arrives.</li>
-         *     <li>{@link #ADDING} if an item is being added to the currently active {@link Accumulator}</li>
-         *     <li>{@link ItemsTerminated} if the items source terminated but the target is not yet terminated.</li>
-         *     <li>{@link #TERMINATED} if the target subscriber has been terminated (or cancelled).</li>
          *     <li>{@link Accumulator} which is emitted by the boundaries source.</li>
+         *     <li>{@link #ADDING} if an item is being added to the currently active {@link Accumulator}.</li>
+         *     <li>{@link NextAccumulatorHolder} if the next boundary is received while still {@link #ADDING} or
+         *     finishing current {@link Accumulator}.</li>
+         *     <li>{@link ItemsTerminated} if the items source terminated but the target is not yet terminated because
+         *     we were delivering onNext or because where was not demand.</li>
+         *     <li>{@link #TERMINATED} if the target subscriber has been terminated (or cancelled).</li>
          * </ul>
          */
         @Nullable
         private volatile Object maybeAccumulator;
+        /**
+         * Number of pending boundaries to deliver to the target based on expressed demand.
+         * <ul>
+         *     <li>Non-negative value - current number of pending accumulations.</li>
+         *     <li>{@link #NEED_TERMINATE} - items terminated but one accumulation is waiting for more demand.</li>
+         *     <li>{@link #NEED_REQUEST_ITEMS} - more demand for accumulations should drive more demand for items.</li>
+         * </ul>
+         */
+        private volatile long pending;
+
+        final int itemsRequestN;
+
+        State(final int itemsRequestN) {
+            this.itemsRequestN = itemsRequestN;
+        }
+
+        <T, B> boolean requested(long n, final Subscriber<? super B> target) {
+            final long oldPending = pendingUpdater.getAndAccumulate(this, n, (prev, nValue) -> {
+                if (nValue <= 0L || prev == NEED_TERMINATE) {
+                    return prev;
+                }
+                return prev == NEED_REQUEST_ITEMS ? nValue : addWithOverflowProtection(prev, nValue);
+            });
+            if (oldPending == NEED_TERMINATE && n > 0L) {
+                @SuppressWarnings("unchecked")
```

Comment:
what if we are in `NEED_TERMINATE` state and someone does `request(1); request(1);`? we should check `ItemsTerminated instance of ItemsTerminated ` to avoid class cast (we should be careful to avoid duplicate termination too).

add an additional `bufferSubscriber.awaitSubscription().request(1);` to `terminalBeforeRequestedDelivers` (line 571) to test this issue.